statsmodels/tools/tools.py

```diff
@@ -233,108 +232,37 @@ def categorical(data, col=None, dictnames=False, drop=False, ):
-            pass
-        else:
-            raise ValueError("Option {0} not understood for "
-                             "has_constant.".format(has_constant))
-    if not prepend:
-        columns = [data.name, 'const']
-    else:
-        columns = ['const', data.name]
-    results = DataFrame({data.name : data, 'const' : const}, columns=columns)
-    return results
-
-
-def _dataframe_add_constant(data, prepend, has_constant):
-    # check for const.
-    if np.any(data.var(0) == 0):
-        if has_constant == 'raise':
-            raise ValueError("data already contains a constant.")
-        elif has_constant == 'skip':
-            return data
-        elif has_constant == 'add':
-            pass
-        else:
-            raise ValueError("Option {0} not understood for "
-                             "has_constant.".format(has_constant))
-    if prepend:
-        data.insert(0, 'const', 1)
-    else:
-        data['const'] = 1
-    return data
-
-
-def _pandas_add_constant(data, prepend, has_constant):
-    from pandas import Series
-    if isinstance(data, Series):
-        return _series_add_constant(data, prepend, has_constant)
-    else:
-        return _dataframe_add_constant(data, prepend, has_constant)
-
-
 # TODO: add an axis argument to this for sysreg
 def add_constant(data, prepend=True, has_constant='skip'):
     """
-    This appends a column of ones to an array if prepend==False.
+    Adds a column of ones to an array
 
     Parameters
     ----------
     data : array-like
         `data` is the column-ordered design matrix
     prepend : bool
-        True and the constant is prepended rather than appended.
+        If true, the constant is in the first column.  Else the constant is
+        appended (last column).
     has_constant : str {'raise', 'add', 'skip'}
-        Behavior if `data` already has a constant. The default will return
+        Behavior if ``data'' already has a constant. The default will return
         data without adding another constant. If 'raise', will raise an
         error if a constant is present. Using 'add' will duplicate the
-        constant, if one is present. Has no effect for structured or
-        recarrays. There is no checking for a constant in this case.
+        constant, if one is present.
 
     Returns
     -------
-    data : array
-        The original array with a constant (column of ones) as the first or
-        last column.
+    data : array, recarray or DataFrame
+        The original values with a constant (column of ones) as the first or
+        last column. Returned value depends on input type.
+
+    Notes
+    -----
+    When the input is recarray or a pandas Series or DataFrame, the added
+    column's name is 'const'.
     """
-    if _is_using_pandas(data, None):
-        # work on a copy
-        return _pandas_add_constant(data.copy(), prepend, has_constant)
-    else:
-        data = np.asarray(data)
-    if not data.dtype.names:
```

Comment:
I think we should also handle the plain numpy array case here. 
It's the most common case, and it's a long way to track the code to where it is done in the current changes, when all we need is essentially a np.column_stack, even with a bit of code duplication.

I would also check for constants only if has_constant != 'add'